File: 2024/13.py
import time, numpy as np

from utils import readInput

class Machine:
    def __init__(self):
        self.buttonA = None
        self.buttonB = None
        self.button = np.zeros(shape=(2, 2))
        self.prize = None

    def solve(self, use_numpy=False):
        if use_numpy:
            sol = np.linalg.solve(self.button.T, self.prize)
            if all(np.abs(np.round(sol) - sol) < 1e-3):
                return sol
            else:
                return None
        else:
            a = round((self.prize[1]-((self.button[1,1]*self.prize[0])/self.button[1,0]))/(self.button[0,1]-((self.button[1,1]*self.button[0,0])/self.button[1,0])))
            b = round((self.prize[0]-self.button[0,0]*a)/self.button[1,0])
            if self.button[0,0]*a+self.button[1,0]*b==self.prize[0] and self.button[0,1]*a+self.button[1,1]*b==self.prize[1]:
                return (a, b)
            else:
                return None

    def cost(self):
        pushes = self.solve()
        if pushes is None:
            return 0
        else:
            return pushes[0]*3 + pushes[1]*1
    
    def __repr__(self):
        return (f"A: {self.buttonA}, B: {self.buttonB}, prize: {self.prize}")

def loadInput(filename: str) -> list[Machine]:
    lines = readInput(filename)
    machines: list[Machine] = []
    for l in lines:
        if l == "":
            continue
        if l.startswith("Button A"):
            machine: Machine = Machine()
            #machine.buttonA = [int(item.split("+")[1]) for item in l.split(",")]
            machine.button[0] = np.array([int(item.split("+")[1]) for item in l.split(",")])
        elif l.startswith("Button B"):
            #machine.buttonB = [int(item.split("+")[1]) for item in l.split(",")]
            machine.button[1] = np.array([int(item.split("+")[1]) for item in l.split(",")])
        elif l.startswith("Prize"):
            machine.prize = [int(item.split("=")[1]) for item in l.split(",")]
            machines.append(machine)
    return machines

def part1(machines: list[Machine]):
    cost = sum([m.cost() for m in machines])
    print (f"🎄 Part 1: {cost}", end='')
    
def part2(machines: list[Machine]):
    const = 10000000000000
    cost = 0
    for m in machines:
        m.prize = [const+p for p in m.prize]
        cost += m.cost()
    print (f"🎄🎅 Part 2: {cost}", end='')

if __name__ == '__main__':
    title = "Day 13: Claw Contraption"
    sub = "❄ "*(len(title)//2+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput("input_13.txt")
    
    t0 = time.time()
    part1(inputs)
    print (" - {:.6f} s".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs)
    print (" - {:.6f} s".format(time.time()-t0))

File: 2024/13.jl

mutable struct Machine
    buttonA::Int
    buttonB::Int
    button::Matrix{Int}
    prize::Vector{Int}
end

function solve(m::Machine)
    a = round((m.prize[2]-((m.button[2,2]*m.prize[1])/m.button[2,1]))/(m.button[1,2]-((m.button[2,2]*m.button[1,1])/m.button[2,1])))
    b = round((m.prize[1]-m.button[1,1]*a)/m.button[2,1])
    if m.button[1,1]*a+m.button[2,1]*b == m.prize[1] && m.button[1,2]*a+m.button[2,2]*b == m.prize[2]
        return (a, b)
    else
        return nothing
    end
end

function cost(m::Machine)
    pushes = solve(m)
    if pushes === nothing
        return 0
    else
        return pushes[1]*3 + pushes[2]*1
    end
end

function toString(m::Machine)
    "A: $(m.buttonA), B: $(m.buttonB), prize: $(m.prize)"
end

function loadInput(filename::String)
    lines = readlines(filename)
    machines = []
    for i in 1:4:length(lines)
        machine = Machine(0, 0, zeros(2,2), [0,0])
        machine.button[1, :] = parse.(Int, map(x -> x[2], split.(split(lines[i], ","), "+")))
        machine.button[2, :] = parse.(Int, map(x -> x[2], split.(split(lines[i+1], ","), "+")))
        machine.prize = parse.(Int, map(x -> x[2], split.(split(lines[i+2], ","), "=")))
        push!(machines, machine)
    end
    machines
end
      
function part1(machines)
    tot_cost = 0
    for m in machines
        tot_cost += cost(m)
    end
    println("🎄 Part 1:  $(tot_cost)")
end

function part2(machines)
    constant = 10000000000000
    tot_cost = 0
    for m in machines
        m.prize = [m.prize[1] + constant, m.prize[2] + constant]
        tot_cost += cost(m)
    end
    println("🎄🎅 Part 2: $(tot_cost)")
end

function main()
    title = "Day 13: Claw Contraption"
    sub = "❄ "^(length(title)÷2 + 2)
    
    println()
    println(" $title ")
    println(sub)
    
    inputs = loadInput("input_13.txt")
    @time part1(inputs)
    @time part2(inputs)
end

main()